libminifi/src/utils/NetworkInterfaceInfo.cpp

```diff
@@ -0,0 +1,169 @@
+  }
+  const std::string& name_;
+};
+}
+
+std::vector<NetworkInterfaceInfo> NetworkInterfaceInfo::getNetworkInterfaceInfos(std::function<bool(const NetworkInterfaceInfo&)> filter,
+    const std::optional<uint32_t> max_interfaces) {
+  std::vector<NetworkInterfaceInfo> network_adapters;
+#ifdef WIN32
+  ULONG buffer_length = sizeof(IP_ADAPTER_ADDRESSES);
+  auto get_adapters_err = GetAdaptersAddresses(0, 0, nullptr, nullptr, &buffer_length);
+  if (ERROR_BUFFER_OVERFLOW != get_adapters_err) {
+    logger_->log_error("GetAdaptersAddresses failed: %lu", get_adapters_err);
+    return network_adapters;
+  }
+  std::vector<char> bytes(buffer_length, 0);
+  IP_ADAPTER_ADDRESSES* adapter = reinterpret_cast<IP_ADAPTER_ADDRESSES*>(bytes.data());
+  get_adapters_err = GetAdaptersAddresses(0, 0, nullptr, adapter, &buffer_length);
+  if (NO_ERROR != get_adapters_err) {
+    logger_->log_error("GetAdaptersAddresses failed: %lu", get_adapters_err);
+    return network_adapters;
+  }
+  while (adapter != nullptr) {
+    NetworkInterfaceInfo interface_info(adapter);
+    if (filter(interface_info)) {
+      auto it = std::find_if(network_adapters.begin(), network_adapters.end(), HasName(interface_info.getName()));
+      if (it == network_adapters.end()) {
+        network_adapters.emplace_back(std::move(interface_info));
+      } else {
+        interface_info.moveAddressesInto(*it);
+      }
+    }
+    if (max_interfaces.has_value() && network_adapters.size() >= max_interfaces.value())
+      return network_adapters;
+    adapter = adapter->Next;
+  }
+#else
+  struct ifaddrs* interface_addresses = nullptr;
+  auto cleanup = gsl::finally([interface_addresses] { freeifaddrs(interface_addresses); });
+  if (getifaddrs(&interface_addresses) == -1) {
+    logger_->log_error("getifaddrs failed: %s", std::strerror(errno));
+    return network_adapters;
+  }
+
+  for (struct ifaddrs* ifa = interface_addresses; ifa != nullptr; ifa = ifa->ifa_next) {
+    if (!ifa->ifa_addr)
+      continue;
+    NetworkInterfaceInfo interface_info(ifa);
+    if (filter(interface_info)) {
+      auto it = std::find_if(network_adapters.begin(), network_adapters.end(), HasName(interface_info.getName()));
+      if (it == network_adapters.end()) {
+        network_adapters.emplace_back(std::move(interface_info));
+      } else {
+        interface_info.moveAddressesInto(*it);
+      }
+    }
+    if (max_interfaces.has_value() && network_adapters.size() >= max_interfaces.value())
+      return network_adapters;
+  }
+#endif
+  return network_adapters;
+}
+
+void move_append(std::vector<std::string>& source, std::vector<std::string>& destination) {
+  if (source.size() == 0)
+    return;
+  if (destination.empty()) {
+    destination = std::move(source);
+  } else {
+    destination.reserve(destination.size() + source.size());
+    std::move(std::begin(source), std::end(source), std::back_inserter(destination));
+    source.clear();
+  }
```

Comment:
The last block would cover the special cases, too. Let's simplify this code!
```suggestion
  destination.reserve(destination.size() + source.size());
  std::move(std::begin(source), std::end(source), std::back_inserter(destination));
  source.clear();
```